### PR TITLE
Add ? to the list of things to quote when it's an exact value.

### LINKED ISSUE #177: lone question mark scalar should be quoted, not encoded in Plain Style
#### Issue description
The encoding of a single `?` character as a value should be quoted, so as not to be confused with the "?" taken as a _mapping key indicator_.

## Example

Given this Elixir map:

```elixir
%{
  "foo" => [
    42,
    "?",
    "bar"
  ]
}
```

Ymlr 5.0 produces this YML:

```yml
---
foo:
  - 42
  - ?
  - bar
```

The original string "?" is being emitted as a plain `?` without quotation. This seems to result in ambiguity, since the "?" character, in some contexts, is also used to indicate a key in a map.

I first discovered this when I found that [`yaml_elixir`](https://hexdocs.pm/yaml_elixir/readme.html), which is a wrapper around `yamerl`, failed to parse the above Ymlr-produced document, saying that it is malformed YML.

Consider the output of running that YML through the parser tests in [libyaml](https://github.com/yaml/libyaml/blob/master/tests/ReadMe.md).

Write the above YML document to `question.yml` and then:

`cat question.yml | ./tests/run-parser-test-suite`

This shows the various parser events, according to libyaml, as it makes its way through the document:

```
+STR
+DOC
+MAP
=VAL :foo
+SEQ
=VAL :42
+MAP
=VAL :
=VAL :
-MAP
=VAL :bar
-SEQ
-MAP
-DOC
-STR
```

In particular, notice what happens after it recognizes `42` as a value in the sequence (`=VAL :42`). The next event is `+MAP`. It starts a new map. And then it emits two events of `=VAL :`, and then it closes that map with `-MAP`.

I take the two `=VAL :` events as an empty key followed by an empty value. This one key/value pair is the only one in that map.

Why would it do that? I think it would do that if it was interpreting that plain `?` question as a mapping key _indicator_. That puts it in "map mode", scanning for a key, after which it will scan for a value, and so on, until there's an event that would close that map: namely, the `- bar`, which is the next item in the outer sequence.

#### Link to documentation

The spec's [Section 5.3 on Indicators](https://yaml.org/spec/1.2.2/#53-indicator-characters) says:

> "?" (`x3F`, question mark) denotes a [mapping key](https://yaml.org/spec/1.2.2/#nodes).

What does this mapping key do? Example 5.3 in the docs illustrates.

CleanShot 2023-12-26 at 13 07 14@2x

In that example, the `?` followed by `space` and then `sky` indicates that `sky` should be treated as a key in a map. The following term `: blue` is taken as the value that goes with the key `sky`.

[Section 7.3.3 on Plain Style](https://yaml.org/spec/1.2.2/#733-plain-style) says:

> Plain scalars must not begin with most [indicators](https://yaml.org/spec/1.2.2/#indicator-characters), as this would cause ambiguity with other YAML constructs. However, the “:”, “?” and “-” [indicators](https://yaml.org/spec/1.2.2/#indicator-characters) may be used as the first character if followed by a non-[space](https://yaml.org/spec/1.2.2/#white-space-characters) “safe” character, as this causes no ambiguity.

Here's how I reason through the implications of that for our case:

Since `?` is an indicator, the only way the line `- ?` could be taken as Plain Style, is if the `?` were _followed by_ a "safe" character.

But since `?` is the entire value, it's not followed by _any_ character, and therefore it's not followed by a _safe_ character. Therefore, the value `?` cannot be encoded in Plain Style; it must be escaped or quoted.

### PR DESCRIPTION
**Pull request by mlwilkerson**
Closes #177 

The linked issue includes the full description of the problem this resolves, with reproduction, and notes from the YAML spec. The particularly relevant sections of the spec are:

- [5.3](https://yaml.org/spec/1.2.2/#53-indicator-characters) - the `?` character is used as an `indicator` character, indicating a mapping key
- [7.3.3](https://yaml.org/spec/1.2.2/#733-plain-style) - the `?` character can only be encoded in Plain Style when it's followed by a non-space "safe" character.

The proposed fix is to add `?` to the list of values already being considered for quotation when they appear as exact values.

---

#### Requirements

- [x] Entry in CHANGELOG.md was created
- [x] Link to documentation on https://yaml.org/ is provided in the PR description
- [x] Functionality is covered by newly created tests

### CODE REVIEW

File: test/ymlr/encode_test.exs [hunk 1]
```diff
@@ -49,6 +49,7 @@ defmodule Ymlr.EncodeTest do
     end
 
     test "quoted strings - starts with special char" do
+      assert_identity_and_output("?", ~S('?'))
       assert_identity_and_output("!tag", ~S('!tag'))
```
Comment (mruoss) [inline test/ymlr/encode_test.exs:53]: Reading the documentation you liked, I think we should also add tests for the other indicator characters (ie. the this list: `[22] c-indicator = `. Also, I would put them in a separate `test` block: Would you add that to your PR? Otherwise we can merge and I'll do it.


```suggestion
    test "quoted strings - indicator characters on their own" do
      assert_identity_and_output("-", ~S('-'))
      assert_identity_and_output("?", ~S('?'))
      assert_identity_and_output(":", ~S(':'))
      assert_identity_and_output("[", ~S('['))
      assert_identity_and_output("]", ~S(']'))
      assert_identity_and_output("{", ~S('{'))
      assert_identity_and_output("}", ~S('}'))
      assert_identity_and_output("#", ~S('#'))
      assert_identity_and_output("&", ~S('&'))
      assert_identity_and_output("&", ~S('&'))
      assert_identity_and_output("*", ~S('*'))
      assert_identity_and_output("*", ~S('*'))
      assert_identity_and_output("!", ~S('!'))
      assert_identity_and_output("!", ~S('!'))
      assert_identity_and_output("|", ~S('|'))
      assert_identity_and_output("|", ~S('|'))
      assert_identity_and_output(">", ~S('>'))
      assert_identity_and_output(">", ~S('>'))
      assert_identity_and_output("%", ~S(%))
      assert_identity_and_output("@", ~S('@'))
      assert_identity_and_output("`", ~S('`'))
    end
    
    test "quoted strings - starts with special char" do
      assert_identity_and_output("!tag", ~S('!tag'))
```

**Review by mruoss — COMMENTED** [review]
Thanks a lot for the PR and for taking the time to look up and link the documentation. Reading it, made me think we should add tests for all indicator chars if they're on their own. See the comment on the diff.

File: test/ymlr/encode_test.exs [hunk 2]
```diff
@@ -18,6 +18,14 @@ defmodule Ymlr.EncodeTest do
...
       assert_identity_and_output("hello world", "hello world")
       assert_identity_and_output("that's it", "that's it")
+
+      # Strings where the ":", "?", or "-" indicator character is used as the
+      # first character may be encoded in Plain Style (without being quoted)
+      # if followed by a non-space “safe” character.
+      # https://yaml.org/spec/1.2.2/#733-plain-style
+      assert_identity_and_output("?x", ~S(?x))
+      assert_identity_and_output(":x", ~S(:x))
+      assert_identity_and_output("-x", ~S(-x))
```
Comment (mlwilkerson) [inline test/ymlr/encode_test.exs:28]: If we're expanding test coverage to handle some of these corner cases pertaining to indicator characters, then I'd also propose adding these tests.

File: lib/ymlr/encode.ex [hunk 3]
```diff
@@ -106,6 +106,9 @@ defmodule Ymlr.Encode do
   @single_quote_when_exact [
     "",
     "~",
+    "?",
+    "-",
```
Comment (mlwilkerson) [inline lib/ymlr/encode.ex:110]: One of your proposed test cases to cover _all_ indicators on their own failed. Adding this to include `-` among those that should be single quoted when exact caused that test to pass.

Comment (mruoss) [reply lib/ymlr/encode.ex:110]: I see. Thanks for the fix!

**Review by mruoss — APPROVED** [review]

**Comment by mruoss** [comment]
Thanks a lot!
